rust_src/src/casetab.rs

```diff
@@ -233,24 +257,51 @@ pub unsafe extern "C" fn get_canonical_case_table() -> LispObject {
 ///  (of characters with the same canonical equivalent); it may be nil,
 ///  in which case it is deduced from CANONICALIZE.
 #[lisp_fn(name = "set-case-table", c_name = "set_case_table")]
-pub fn set_case_table_lisp(table: LispObject) -> LispObject {
-    set_case_table(table, false)
+pub fn set_case_table_lisp(table: LispObject) -> LispCaseTable {
+    let mut case_table = LispCaseTable::from(table);
+    if !case_table.is_proper() {
+        wrong_type!(Qcase_table, table);
+    }
+    case_table.setup_extras();
+
+    let mut buffer: LispBufferRef = current_buffer().into();
+    buffer.set_case_table(case_table);
+
+    case_table
 }
 
 /// Select a new standard case table for new buffers.
 /// See `set-case-table' for more info on case tables.
 #[lisp_fn]
-pub fn set_standard_case_table(table: LispObject) -> LispObject {
-    set_case_table(table, true)
+pub fn set_standard_case_table(table: LispObject) -> LispCaseTable {
+    let case_table = LispCaseTable::from(table);
+    if !case_table.is_proper() {
+        wrong_type!(Qcase_table, table);
+    }
+    set_standard_case_table_1(case_table);
+    case_table
+}
+
+fn set_standard_case_table_1(mut case_table: LispCaseTable) {
```

Comment:
Making `is_proper` required for conversion using `From` would remove the need for two functions here.